collector/zfs.go

```diff
@@ -0,0 +1,157 @@
+package collector
+
+// +build !nofilesystem
+// +build !nozfs
+
+import (
+	"errors"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+)
+
+type zfsMetricValue int
+
+const zfsErrorValue = zfsMetricValue(-1)
+
+var zfsNotAvailableError = errors.New("ZFS / ZFS statistics are not available")
+
+type zfsSysctl string
+type zfsSubsystemName string
+
+const (
+	arc = zfsSubsystemName("zfs_arc")
+)
+
+//------------------------------------------------------------------------------
+//                                    Metrics
+//------------------------------------------------------------------------------
+
+type zfsMetric struct {
+	subsystem zfsSubsystemName // The Prometheus subsystem name
+	name      string           // The Prometheus name of the metric
+	sysctl    zfsSysctl        // The sysctl of the ZFS metric
+}
+
+func NewZFSMetric(subsystem zfsSubsystemName, sysctl, name string) zfsMetric {
+	return zfsMetric{
+		sysctl:    zfsSysctl(sysctl),
+		subsystem: subsystem,
+		name:      name,
+	}
+}
+func (m *zfsMetric) BuildFQName() string {
+	return prometheus.BuildFQName(Namespace, string(m.subsystem), m.name)
+}
+
+func (m *zfsMetric) HelpString() string {
+	return m.name
+}
+
+//------------------------------------------------------------------------------
+//                                 Collector
+//------------------------------------------------------------------------------
+
+func init() {
+	Factories["zfs"] = NewZFSCollector
+}
+
+type zfsCollector struct {
+	zfsMetrics     []zfsMetric
+	metricProvider zfsMetricProvider
+}
+
+func NewZFSCollector() (Collector, error) {
+	err := zfsInitialize()
+	switch {
+	case err == zfsNotAvailableError:
+		log.Debug(err)
+		break
+		return &zfsCollector{}, err
+	}
+
+	return &zfsCollector{
+		zfsMetrics: []zfsMetric{
+			NewZFSMetric(arc, "kstat.zfs.misc.arcstats.p", "mru_size"),
+			NewZFSMetric(arc, "kstat.zfs.misc.arcstats.size", "size"),
+			NewZFSMetric(arc, "kstat.zfs.misc.arcstats.c_min", "target_min_size"),
+			NewZFSMetric(arc, "kstat.zfs.misc.arcstats.c", "target_size"),
+			NewZFSMetric(arc, "kstat.zfs.misc.arcstats.c_max", "target_max_size"),
+			NewZFSMetric(arc, "kstat.zfs.misc.arcstats.hits", "hits"),
+			NewZFSMetric(arc, "kstat.zfs.misc.arcstats.misses", "misses"),
+		},
+		metricProvider: NewZFSMetricProvider(),
+	}, nil
+}
+
+func (c *zfsCollector) Update(ch chan<- prometheus.Metric) (err error) {
+
+	log.Debug("Preparing metrics update")
+	err = c.metricProvider.PrepareUpdate()
```

Comment:
You haven't changed this, you should have this return the map directly rather than going via an object.

You should also be able to merge the two classes you currently have.